utils/progress_bar.go

```diff
@@ -0,0 +1,122 @@
+package utils
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/mendersoftware/log"
+	"golang.org/x/sys/unix"
+)
+
+type Units int
+
+const (
+	BYTES Units = iota
+	TICKS
+	NONE
+)
+
+const (
+	defaultTerminalWidth = 32
+)
+
+type ProgressBar struct {
+	out        *os.File
+	N          uint64
+	c          uint64
+	pchar      string
+	prefix     string
+	units      Units
+	exceeded   bool
+	isTerminal bool
+}
+
+func NewProgressBar(out *os.File, N uint64, units Units) *ProgressBar {
+	isTerminal := true
+	_, err := unix.IoctlGetTermios(int(out.Fd()), unix.TCGETS)
+	if err != nil {
+		isTerminal = false
+	}
+	return &ProgressBar{
+		out:        out,
+		N:          N,
+		c:          0,
+		pchar:      "#",
+		units:      units,
+		exceeded:   false,
+		isTerminal: isTerminal,
+	}
+}
+
+func (pb *ProgressBar) SetProgressChar(char rune) {
+	pb.pchar = string(char)
+}
+
+func (pb *ProgressBar) SetPrefix(prefix string) {
+	pb.prefix = prefix
+}
+
+func (pb *ProgressBar) Tick(n uint64) error {
+
+	var width int
+	var suffix string
+	pb.c += n
+	if pb.c > pb.N {
+		log.Warnf("Progressbar exceeded maximum (%d > %d)", pb.c, pb.N)
+	}
+	percent := float64(pb.c) / float64(pb.N)
+
+	if pb.isTerminal {
+		// Adjust to terminal width
+		winSz, err := unix.IoctlGetWinsize(int(pb.out.Fd()), unix.TIOCGWINSZ)
+		if err != nil {
+			return err
+		}
+		width = int(winSz.Col)
+		if _, err := pb.out.WriteString("\r"); err != nil {
+			return err
+		}
+	} else {
+		width = defaultTerminalWidth
+	}
+
+	switch pb.units {
+	case BYTES:
+		suffix = ShortSize(pb.c)
+		suffix = fmt.Sprintf("%3d%% | %s", int(100*percent), suffix)
+	case TICKS:
+		suffix = fmt.Sprintf("%3d%% | #%d", int(100*percent), pb.c)
+	default:
+		// None
+		suffix = ""
+	}
+
+	progWidth := width - len(suffix) - len(pb.prefix) - 2
+	if progWidth <= 0 {
+		// Ignore nasty line wrapping and force default
+		progWidth = defaultTerminalWidth
+	}
+
+	numPChars := int(float64(progWidth) * percent)
+	pChars := strings.Repeat(pb.pchar, numPChars)
+	spaces := strings.Repeat(" ", progWidth-numPChars)
+	_, err := pb.out.WriteString(fmt.Sprintf(
+		"%s[%s%s]%s", pb.prefix, pChars, spaces, suffix))
+	return err
+}
+
+// TODO: Create a separate utility source file with funcs like this
+func ShortSize(bytes uint64) string {
+	var suffixes = [...]string{"B", "KB", "MB", "GB", "TB", "UWOT?!"}
```

Comment:
Hehe, I would stop at `TB`. Also, for complete inambiguity, I would use `"KiB", "MiB", "GiB", "TiB"`, since these are in fact [Kibibytes](https://en.wikipedia.org/wiki/Kibibyte), etc. Yes, the name is silly, but it is the only unambiguous option. :-)